### PR TITLE
Rewrite peers module - #409

### PR DESCRIPTION
**Pull request by 4miners**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "semver": "=5.3.0",
     "socket.io": "=1.4.8",
     "sodium": "LiskHQ/node-sodium#07ba174",
+    "steed": "=1.1.3",
```
Comment (4miners) [inline package.json]: Not needed.

File: app.js [hunk 2]
```diff
@@ -445,8 +449,9 @@ d.run(function () {
 			});
 		}],
 
-		ready: ['modules', 'bus', function (scope, cb) {
+		ready: ['modules', 'bus', 'logic', function (scope, cb) {
 			scope.bus.message('bind', scope.modules);
+			scope.logic.peers.bind(scope);
```
Comment (4miners) [inline app.js]: Expose modules and library to `logic.peers`

File: modules/peers.js [hunk 3]
```diff
@@ -1,35 +1,29 @@
...
 var ip = require('ip');
 var OrderBy = require('../helpers/orderBy.js');
 var path = require('path');
-var Peer = require('../logic/peer.js');
-var PeerSweeper = require('../logic/peerSweeper.js');
+var pgp = require('pg-promise')(); // We also initialize library here
 var Router = require('../helpers/router.js');
 var sandboxHelper = require('../helpers/sandbox.js');
 var schema = require('../schema/peers.js');
+var async = require('async');
```
Comment (4miners) [inline modules/peers.js]: Should be sorted.

File: modules/peers.js [hunk 4]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
 };
 
-__private.banManager = function (cb) {
-	library.db.query(sql.banManager, { now: Date.now() }).then(function (res) {
-		return setImmediate(cb, null, res);
+__private.dbLoad = function (cb) {
+	var updated = 0;
+	library.logger.trace('Importing peers from database');
+	library.db.any(sql.getAll).then(function (rows) {
+		library.logger.trace('Imported peers from database', {count: rows.length});
```
Comment (4miners) [inline modules/peers.js]: Should be `info` level

File: modules/peers.js [hunk 5]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
+		});
 
-	if (filter.state >= 0) {
-		where.push('"state" = ${state}');
-		params.state = filter.state;
-	}
+		return t.batch(queries);
+	})
+	.then(function (data) {
+		library.logger.debug('Peers exported to database');
```
Comment (4miners) [inline modules/peers.js]: Should be `info` level

File: modules/peers.js [hunk 6]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
 
-	if (filter.broadhash) {
-		where.push('"broadhash" = ${broadhash}');
-		params.broadhash = filter.broadhash;
+Peers.prototype.remove = function (pip, port) {
+	var frozenPeer = _.find(library.config.peers.list, function (peer) {
+		return peer.ip === pip && peer.port === port;
+	});
+	if (frozenPeer) {
+		//FIXME: Keeping peer frozen is bad idea at all
```
Comment (4miners) [inline modules/peers.js]: I'm still not sure here

File: modules/peers.js [hunk 7]
```diff
@@ -297,135 +370,73 @@ Peers.prototype.acceptable = function (peers) {
...
 				return randomList(options, peers, waterCb);
 			} else {
 				return setImmediate(waterCb, null, peers);
 			}
 		}
 	], function (err, peers) {
+		// Calculate consensus
 		var consensus = Math.round(options.matched / peers.length * 100 * 1e2) / 1e2;
-		    consensus = isNaN(consensus) ? 0 : consensus;
+			consensus = isNaN(consensus) ? 0 : consensus;
```
Comment (4miners) [inline modules/peers.js]: intend

File: modules/transport.js [hunk 8]
```diff
@@ -45,17 +45,16 @@ __private.attachApi = function () {
...
 				ip: req.ip,
 				port: req.headers.port
 			}
 		);
 
-		var headers = req.peer.extend(req.headers);
+		var headers = req.peer.applyHeaders(req.headers);
 
 		req.sanitize(headers, schema.headers, function (err, report) {
-			if (err) { return next(err); }
```
Comment (4miners) [inline modules/transport.js]: We will never have `err` here, so no need to check for

File: modules/transport.js [hunk 9]
```diff
@@ -245,6 +244,12 @@ __private.attachApi = function () {
 		});
 	});
 
+	router.get('/ping', function (req, res) {
```
Comment (4miners) [inline modules/transport.js]: New endpoint, experimental ;)

File: modules/transport.js [hunk 10]
```diff
@@ -348,12 +353,12 @@ __private.banPeer = function (options) {
 		library.logger.trace('Peer ban skipped', {options: options});
 		return false;
 	}
-	library.logger.warn([options.code, ['Ban', options.peer.string, (options.clock / 60), 'minutes'].join(' '), options.req.method, options.req.url].join(' '));
-	modules.peers.state(options.peer.ip, options.peer.port, 0, options.clock);
+	library.logger.debug([options.code, ['Ban', options.peer.string, (options.clock / 60), 'minutes'].join(' '), options.req.method, options.req.url].join(' '));
```
Comment (4miners) [inline modules/transport.js]: Changeg to `debug` level because it's not so important

File: modules/transport.js [hunk 11]
```diff
@@ -348,12 +353,12 @@ __private.banPeer = function (options) {
...
+	modules.peers.ban(options.peer.ip, options.peer.port, options.clock);
 };
 
 __private.removePeer = function (options) {
-	library.logger.warn([options.code, 'Removing peer', options.peer.string, options.req.method, options.req.url].join(' '));
+	library.logger.debug([options.code, 'Removing peer', options.peer.string, options.req.method, options.req.url].join(' '));
```
Comment (4miners) [inline modules/transport.js]: Changeg to `debug` level because it's not so important

File: modules/transport.js [hunk 12]
```diff
@@ -516,12 +521,12 @@ Transport.prototype.getFromPeer = function (peer, options, cb) {
 		url = options.url;
 	}
 
-	peer = modules.peers.accept(peer);
+	peer = library.logic.peers.create(peer);
 
 	var req = {
 		url: 'http://' + peer.ip + ':' + peer.port + url,
 		method: options.method,
-		headers: extend({}, __private.headers, options.headers),
```
Comment (4miners) [inline modules/transport.js]: No need extend here, because we will never have `options.headers`

File: schema/peers.js [hunk 13]
```diff
@@ -51,7 +51,7 @@ module.exports = {
 					minimum: 1
 				}
 			},
-			required: ['ip', 'port', 'state']
```
Comment (4miners) [inline schema/peers.js]: We only need `ip` and `port` for discover peers

File: schema/transport.js [hunk 14]
```diff
@@ -41,7 +37,7 @@ module.exports = {
 				minimum: 1
 			}
 		},
-		required: ['ip', 'port', 'version', 'nethash']
```
Comment (4miners) [inline schema/transport.js]: `ip` is not header, so removed from here

File: logic/peer.js [hunk 15]
```diff
@@ -36,25 +51,35 @@ Peer.prototype.accept = function (peer) {
...
 
-	this.port = this.parseInt(peer.port, 0);
+	this.port = self.parseInt(peer.port, 0);
 
-	if (this.ip) {
-		this.string = (this.ip + ':' + this.port || 'unknown');
-	} else {
-		this.string = 'unknown';
+	if (this.ip && this.port) {
+		this.string = this.ip + ':' + this.port;
```
Comment (MaciejBaj) [inline logic/peer.js]: string? maybe address better?

File: logic/peers.js [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+	}
+};
+
+Peers.prototype.exists = function (peer) {
+	peer = self.create(peer);
+	return !!__private.peers[peer.string];
+};
+
+Peers.prototype.get = function (peer) {
+	if (typeof peer === 'string') {
```
Comment (MaciejBaj) [inline logic/peers.js]: maybe create some function  'isAddress(peer)' with regexp that will check if it's address or not.

Or maybe split this into 2 functions get and getByAddress

File: logic/peers.js [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	return peer;
+};
+
+Peers.prototype.remove = function (peer) {
+	peer = self.create(peer);
+	// Remove peer if exists
+	if (self.exists(peer)) {
+		library.logger.info('Removed peer', peer.string);
+		library.logger.debug('Removed peer', {peer: __private.peers[peer.string]});
+		__private.peers[peer.string] = null; // Possible memory leak prevention
```
Comment (MaciejBaj) [inline logic/peers.js]: based on http://bertanguven.com/preventing-memory-leaks-in-javascript-null-vs-delete?

File: logic/peers.js [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+	} else {
+		return Object.keys(__private.peers).map(function (key) { return __private.peers[key]; });
+	}
+	
+};
+
+// Public methods
+Peers.prototype.bind = function (scope) {
+	library = scope;
+	modules = scope.modules;
```
Comment (MaciejBaj) [inline logic/peers.js]: doubled?

Comment (MaciejBaj) [inline logic/peers.js]: to keep the logic modules consistent it is better to move the scope (`library = scope`) initialization into constructor and assign here only modules to modules

File: logic/peer.js [hunk 19]
```diff
@@ -25,7 +38,9 @@ Peer.prototype.nullable = [
 	'version',
 	'dappid',
 	'broadhash',
-	'height'
+	'height',
+	'clock',
+	'updated'
 ];
 
 // Public methods
```
Comment (MaciejBaj) [inline logic/peer.js]: the accept function works fine (accepts) peer even without port and ip. Is it ok?

File: logic/peer.js [hunk 20]
```diff
@@ -36,25 +51,35 @@ Peer.prototype.accept = function (peer) {
...
+		}
+	}
+
+	if (peer.height) {
+		this.height = self.parseInt(peer.height, 1);
+	}
+
+	if (peer.clock) {
+		this.clock = peer.clock;
 	}
```
Comment (MaciejBaj) [inline logic/peer.js]: what with `os`, `version`?

File: logic/peer.js (hunk 15 above)
Comment (MaciejBaj) [inline logic/peer.js]: also this.string should be nullified in other case- when proper port and ip are passed first and then undefined values after that `string` remains unchanged

File: logic/peers.js [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+		return Object.keys(__private.peers).map(function (key) { return __private.peers[key].object(); });
+	} else {
+		return Object.keys(__private.peers).map(function (key) { return __private.peers[key]; });
+	}
+	
+};
+
+// Public methods
+Peers.prototype.bind = function (scope) {
+	modules = scope.modules;
```
Comment (MaciejBaj) [inline logic/peers.js]: also modules are not used anywhere here as far

File: logic/peers.js [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+		}
+	});
+
+	library.logger.trace('Peer stats', {total: cnt_total, alive: cnt_active, empty_height: cnt_empty_height, empty_broadhash: cnt_empty_broadhash});
+
+	return true;
+};
+
+Peers.prototype.ban = function (ip, port, seconds) {
+	return library.logic.peers.upsert({
```
Comment (MaciejBaj) [inline logic/peers.js]: self

File: logic/peers.js [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+			++cnt_empty_broadhash;
+		}
+	});
+
+	library.logger.trace('Peer stats', {total: cnt_total, alive: cnt_active, empty_height: cnt_empty_height, empty_broadhash: cnt_empty_broadhash});
+
+	return true;
+};
+
+Peers.prototype.ban = function (ip, port, seconds) {
```
Comment (MaciejBaj) [inline logic/peers.js]: we pass whole peer object / instance to every other function

File: logic/peer.js (hunk 19 above)
Comment (4miners) [inline logic/peer.js]: Yes, that is expected behavior.

File: logic/peer.js (hunk 20 above)
Comment (4miners) [inline logic/peer.js]: Nothing, those values can be set only via `applyHeaders`, but I will re-consider that.

File: logic/peers.js (hunk 16 above)
Comment (4miners) [inline logic/peers.js]: Good idea. Btw, that feature is for flexibility, not used yet.

File: logic/peers.js (hunk 23 above)
Comment (4miners) [inline logic/peers.js]: Yep, logic should be changed here.

File: logic/peers.js [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+
+		if (Object.keys(diff).length) {
+			library.logger.debug('Updated peer ' + peer.string, diff);
+		} else {
+			library.logger.trace('Peer not changed', peer.string);
+		}
+	};
+
+	peer = self.create(peer);
+
```
Comment (4miners) [inline logic/peers.js]: We should check here if peer is valid one (have `string` property).

File: logic/peers.js (hunk 17 above)
Comment (4miners) [inline logic/peers.js]: Based on @Isabello comments and quick research. :)

File: logic/peers.js (hunk 21 above)
Comment (4miners) [inline logic/peers.js]: Yes, I left support for that for future usage. Maybe we should move some other functions from `modules/peers.js` to logic.

File: modules/peers.js [hunk 25]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
+	};
+
+	// Apply filters (by AND)
+	var peers = library.logic.peers.list(true);
+	peers = peers.filter(function (peer) {
+		// var peer = __private.peers[index];
+		var passed = true;
+		_.each(filter, function (value, key) {
+			// Special case for dapp peers
+			if (key === 'dappid' && (peer[key] === null || (Array.isArray(peer[key]) && !_.includes(peer[key], String(value))))) {
```
Comment (4miners) [inline modules/peers.js]: There is something wrong with that, need more tests.

File: modules/peers.js [hunk 26]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
-		params.port = filter.port;
-	}
+	// Creating set of columns
+	var cs = new pgp.helpers.ColumnSet([
+		'ip', 'port',
+		{name: 'state',     def: 1},
+		{name: 'height',    def: 1},
+		{name: 'os',        def: null},
+		{name: 'version',   def: null},
+		{name: 'broadhash', def: null, init: function (col) {
```
Comment (4miners) [inline modules/peers.js]: Need more tests.

File: modules/peers.js [hunk 27]
```diff
@@ -62,232 +56,311 @@ __private.attachApi = function () {
...
+		library.logger.error('Import peers from database failed', {error: err.message || err});
+		return setImmediate(cb);
 	});
 };
 
-__private.getByFilter = function (filter, cb) {
-	var where = [];
-	var params = {};
+__private.dbSave = function (cb) {
+	var peers = library.logic.peers.list();
```
Comment (4miners) [inline modules/peers.js]: We should call normalized version `library.logic.peers.list(true);`, so later not need to set default values for SQL query - will be more readable.

**Comment by karmacoma** [comment]
Closing and reopening with branch conflicts resolved against latest development.

**Pull request closed without merging**